rosbag2_transport/src/rosbag2_transport/rosbag2_transport_python.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Python.h>
+#include <string>
+#include <vector>
+
+#include "rosbag2_transport/rosbag2_transport.hpp"
+#include "rosbag2_transport/record_options.hpp"
+#include "rosbag2_transport/storage_options.hpp"
+
+static PyObject *
+rosbag2_transport_record(PyObject * Py_UNUSED(self), PyObject * args, PyObject * kwargs)
+{
+  rosbag2_transport::StorageOptions storage_options{};
+  rosbag2_transport::RecordOptions record_options{};
+
+  static const char * kwlist[] = {"uri", "storage_id", "all", "topics", nullptr};
+
+  char * uri = nullptr;
+  char * storage_id = nullptr;
+  bool all = false;
+  PyObject * topics = nullptr;
+  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "ss|bO", const_cast<char **>(kwlist),
+    &uri,
+    &storage_id,
+    &all,
+    &topics))
+  {
+    return nullptr;
+  }
+
+  storage_options.uri = std::string(uri);
+  storage_options.storage_id = std::string(storage_id);
+  record_options.all = all;
+
+  if (topics) {
+    PyObject * topic_iterator = PyObject_GetIter(topics);
+    if (topic_iterator != nullptr) {
+      PyObject * topic;
+      while ((topic = PyIter_Next(topic_iterator))) {
+        record_options.topics.emplace_back(PyUnicode_AsUTF8(topic));
+
+        Py_DECREF(topic);
+      }
+      Py_DECREF(topic_iterator);
+    }
+  }
+
+  rosbag2_transport::Rosbag2Transport transport;
+  transport.init();
```

Comment:
can we put this `init()` into the constructor and `shutdown` into the destructor?

Reply:
No, because of testing. `init()` calls `rclcpp::init()` which may only be called once. In tests we need to be able to call `rclcpp::init()` ourselfes. Same reason for `shutdown`.

Reply:
wouldn't it make sense to then constantly require `rclcpp::init()` to be called externally?